package/MDAnalysis/coordinates/H5MD.py

```diff
@@ -0,0 +1,587 @@
+        if 'mpio' is passed for the driver.
+
+        """
+        if self._driver != 'mpio':
+            self.close()
+            self.open_trajectory()
+        else:
+            self._frame = -2
+            self._read_next_timestep()
+
+    @property
+    def has_positions(self):
+        """True if 'position' group is in trajectory."""
+        return self._has['position']
+
+    @has_positions.setter
+    def has_positions(self, value: bool):
+        self._has['position'] = value
+
+    @property
+    def has_velocities(self):
+        """True if 'velocity' group is in trajectory."""
+        return self._has['velocity']
+
+    @has_velocities.setter
+    def has_velocities(self, value: bool):
+        self._has['velocity'] = value
+
+    @property
+    def has_forces(self):
+        """True if 'force' group is in trajectory."""
+        return self._has['force']
+
+    @has_forces.setter
+    def has_forces(self, value: bool):
+        self._has['force'] = value
+
+    @property
+    def n_frames(self):
+        """number of frames in trajectory"""
+        for name, value in self._has.items():
+            if value:
+                return self._particle_group[name]['value'].shape[0]
+
+    def _read_frame(self, frame):
+        """reads data from h5md file and copies to current timestep"""
+        try:
+            for name, value in self._has.items():
+                if value:
+                    myframe = self._particle_group[name]['step'][frame]
+                    break
+            else:
+                raise NoDataError("Provide at least a position, velocity"
+                                  " or force group in the h5md file.")
+        except ValueError:
+            raise IOError from None
+
+        self._frame = frame
+        ts = self.ts
+        particle_group = self._particle_group
+        ts.frame = frame
+
+        # fills data dictionary from 'observables' group
+        # Note: dt is not read into data as it is not decided whether
+        # Timestep should have a dt attribute (see Issue #2825)
+        self._copy_to_data()
+
+
+        # Sets frame box dimensions
+        # Note: H5MD files must contain 'box' group in each particle group
+        if 'edges' in particle_group['box']:
+            ts._unitcell[:] = particle_group['box/edges/value'][frame, :]
+        else:
+            # sets ts.dimensions = [0, 0, 0, 0, 0, 0]
+            ts._unitcell[:] = None
+
+        # set the timestep positions, velocities, and forces with
+        # current frame dataset
+        if self._has['position']:
+            ts.positions = self._get_frame_dataset('position')
+        if self._has['velocity']:
+            ts.velocities = self._get_frame_dataset('velocity')
+        if self._has['force']:
+            ts.forces = self._get_frame_dataset('force')
+
+        if self.convert_units:
+            self._convert_units()
+
+        return ts
+
+    def _read_next_timestep(self):
```

Comment:
Move down and have all the helper methods (`_get_frame_dataset` etc) come right after `_read_frame()` so that they are all in one place.